docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+    from megatron.core.dist_checkpointing.strategies.fully_parallel import FullyParallelLoadStrategyWrapper, FullyParallelSaveStrategyWrapper
+
+    base_save_strategy = TorchDistSaveShardedStrategy('torch_dist', 1)
+    base_load_strategy = TorchDistLoadShardedStrategy()
+
+    def basic_save_load(sharded_state_dict, ckpt_dir):
+        """ Save and load using some basic strategies. """
+        dist_checkpointing.save(sharded_state_dict, ckpt_dir, base_save_strategy)
+        return dist_checkpointing.load(sharded_state_dict, ckpt_dir, base_load_strategy)
+
+
+    def fully_parallel_save_load(sharded_state_dict):
+        """ Save and load using basic strategies wrapped with parallelization strategies. """
+        fully_parallel_save_strategy = FullyParallelSaveStrategyWrapper(base_save_strategy)
+        fully_parallel_load_strategy = FullyParallelLoadStrategyWrapper(base_load_strategy)
+        dist_checkpointing.save(sharded_state_dict, ckpt_dir, fully_parallel_save_strategy)
+        return dist_checkpointing.load(sharded_state_dict, ckpt_dir, fully_parallel_load_strategy)
+
+
+The `dist_checkpointing` package provides default strategies for some sharded backends, so it's enough to specify a tuple `(backend, version)` as a saving strategy.
+Backends and versions are stored in a `metadata.json` file inside the checkpoint so that the loading strategy can be determined automatically (provided that there exists a default loading strategy for a given backend and version).
+
+For "sharded" strategies, currently the backends supported by default are based on `torch.distributed.checkpoint` format (`torch_dist` backend) and Zarr format (`zarr` backend).
+Additionally, as shown in the example above, some wrappers are provided that allow to parallelize the save and load across the whole workload (assuming some data duplication).
+
+For "common" strategies, currently the only supported one is `torch` which saves "common" data into a `common.pt` file.
+
+Optimizers
+==========
+This module gives helper tools to the user to simplify constructing ShardedTensors for optimizer states.
+The ShardedTensors that define local to sharded tensors mapping for model parameters should be reused for optimizer states to avoid code duplication.
+
+To this end, the `dist_checkpointing.optimizers.get_param_id_to_sharded_param_map` function can build a mapping between optimizer params ids and model ShardedTensors.
+This mapping can be used by the `dist_checkpointing.optimizers.optim_state_to_sharding_state` function or application code (for non-standard use cases) to construct optimizer sharded state dict with ShardedTensors.
+This should support most optimizer cases, but some of them might require custom sharded state dict creation.
+A good example is a Distributed Optimizer which flattens the parameters - see `Tensors transformations`_ section for more details.
```

Comment:
The optimizer module provides helpful tools to simplify the construction of ShardedTensors for optimizer states. When defining the local-to-sharded tensor mapping for model parameters, it is recommended that you reuse the same ShardedTensors for optimizer states to avoid code duplication.

To achieve this, the dist_checkpointing.optimizers.get_param_id_to_sharded_param_map function creates a mapping between optimizer parameter IDs and model ShardedTensors. This mapping can be utilized either through the dist_checkpointing.optimizers.optim_state_to_sharding_state function or by writing custom application code (for non-standard use cases) to construct the optimizer sharded state dictionary with ShardedTensors.

While this approach should cover most optimizer scenarios, some may require a custom sharded state dictionary. For instance, consider a Distributed Optimizer that flattens the parameters (refer to the Tensors transformations section for more details).